j5/boards/base.py

```diff
@@ -114,21 +113,16 @@ def __next__(self) -> Board:
         """Get the next item in the iteration."""
         if self._iterator_counter >= len(self.boards):
             raise StopIteration
-        board = self.boards[self._iterator_counter]
+        board = self.boards[list(self.boards.keys())[self._iterator_counter]]
         self._iterator_counter += 1
         return board
 
-    def __getitem__(self, index: BoardIndex) -> Board:
+    def __getitem__(self, serial: str):
         """Get the board from an index."""
         if len(self.boards) <= 0:
             raise IndexError("Could not find any boards.")
-
-        if type(index) == int:
-            return self.boards[cast(int, index)]
-        elif type(index) == str:
-            for b in self.boards:
-                if b.serial == index:
-                    return b
-            raise KeyError(f"Could not find a board with the serial {index}")
-        else:
-            raise IndexError(f"Cannot index boards with type {str(type(index))}")
+        if type(serial) != str:
```

Comment:
Do we need to validate that serial is a `str`?

Reply:
Yes, otherwise the user may use a integer(for example). And instead of receiving a TypeError letting him know that he can only use strings, he would see the KeyError with the message "Could not find a board with the serial {serial}"

Reply:
I would think the user would realise that they need to provide a serial with that error message as I think serials are typically strings. 

I do get your point with having the different exception though, is it worth moving this into the try-except clause? Where when we get the KeyError we then check if its the correct type and raise TypeError there?

Reply:
Remember that we are trying to make the error messages as obvious as possible. Users of this API may have little knowledge of python.

Reply:
Would it really be necessary to move it into the except clause? I feel that it is more readable to be left out of it than perform the if statement inside the except. 
I can do it if you guys feel that it would be the better option.

Reply:
Ok, so I'm just going to make a decision on this one. Both solutions work, but @Skybound1's suggestions are more pythonic, so let's go with that.